cloud/db/environments.ts

```diff
@@ -0,0 +1,515 @@
+      delete: ["ADMIN"],
+    };
+  }
+
+  // ---------------------------------------------------------------------------
+  // Role Resolution
+  // ---------------------------------------------------------------------------
+
+  /**
+   * Determines the user's effective role for an environment.
+   *
+   * Delegates to `ProjectMemberships.getRole` which handles:
+   * - Org OWNER → treated as project ADMIN
+   * - Org ADMIN → treated as project ADMIN
+   * - Explicit project membership role
+   * - No access → NotFoundError (hides environment existence)
+   */
+  getRole({
+    userId,
+    organizationId,
+    projectId,
+  }: {
+    userId: string;
+    organizationId: string;
+    projectId: string;
+    environmentId?: string;
+  }): Effect.Effect<ProjectRole, NotFoundError | DatabaseError, DrizzleORM> {
+    return this.projectMemberships.getRole({
+      userId,
+      organizationId,
+      projectId,
+    });
+  }
+
+  // ---------------------------------------------------------------------------
+  // CRUD Operations
+  // ---------------------------------------------------------------------------
+
+  /**
+   * Creates a new environment within a project.
+   *
+   * Requires ADMIN or DEVELOPER role on the project.
+   *
+   * @param args.userId - The user creating the environment
+   * @param args.organizationId - The organization containing the project
+   * @param args.projectId - The project to create the environment in
+   * @param args.data - Environment data (name)
+   * @returns The created environment
+   * @throws PermissionDeniedError - If user lacks create permission
+   * @throws AlreadyExistsError - If an environment with this name exists in the project
+   * @throws NotFoundError - If the project doesn't exist or user lacks access
+   * @throws DatabaseError - If the database operation fails
+   */
+  create({
+    userId,
+    organizationId,
+    projectId,
+    data,
+  }: {
+    userId: string;
+    organizationId: string;
+    projectId: string;
+    data: Pick<NewEnvironment, "name">;
+  }): Effect.Effect<
+    PublicEnvironment,
+    AlreadyExistsError | NotFoundError | PermissionDeniedError | DatabaseError,
+    DrizzleORM
+  > {
+    return Effect.gen(this, function* () {
+      const client = yield* DrizzleORM;
+
+      yield* this.authorize({
+        userId,
+        action: "create",
+        organizationId,
+        projectId,
+        environmentId: "", // Not used for create
+      });
+
+      // Insert the environment
+      const [environment] = yield* client
+        .insert(environments)
+        .values({
+          name: data.name,
+          projectId,
+        })
+        .returning(publicFields)
+        .pipe(
+          Effect.mapError((e): AlreadyExistsError | DatabaseError =>
+            isUniqueConstraintError(e.cause)
+              ? new AlreadyExistsError({
+                  message: `An environment with name "${data.name}" already exists in this project`,
```

Comment:
If we're enforcing name uniqueness, let's apply the same normalization logic as downstack. I think where we have unique names we should make them slug-compatible names